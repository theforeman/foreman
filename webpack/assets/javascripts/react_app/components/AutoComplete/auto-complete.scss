@import "~patternfly/dist/sass/patternfly/_color-variables.scss";
@import "~patternfly-react/dist/sass/_type-ahead-select.scss";

.foreman-autocomplete {
  .autocomplete-aux {
    position: absolute;
    right: 25px;
    top: 4px;
    z-index: 10;

    .autocomplete-clear-button {
      margin-right: 2px;
      font-size: 12px;
      color: $color-pf-black-600;

      &:hover {
        color: $color-pf-blue-500;
        text-decoration: none;
      }
    }
  }

  .autocomplete-focus-shortcut {
    position: absolute;
    top: 4px;
    left: 4px;
    padding: 0 4px;
    user-select: none;
    font-size: 12px;
    font-weight: bold;
    z-index: 10;
    border-radius: 5px;
    width: 15px;
    height: 19px;
    text-align: center;
    line-height: 17px;
    border: 1px solid rgba($color-pf-black-600, 0.5);
    color: $color-pf-black-600;

    .hide {
      display: none;
    }
  }

  .autocomplete-error {
    position: absolute;
    color: $color-pf-red-200;
  }
}

// The span that contains the loading spinner should move left
// to avoid being override by the clear button.
.rbt-aux {
  right: 12px;
  z-index: 10;
}

.rbt-menu {
  width: auto;
  min-width: -webkit-fill-available !important;
}
