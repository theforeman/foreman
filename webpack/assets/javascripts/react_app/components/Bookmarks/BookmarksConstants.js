export const BOOKMARKS = 'BOOKMARKS';
export const BOOKMARKS_REQUEST = 'BOOKMARKS_REQUEST';
export const BOOKMARKS_SUCCESS = 'BOOKMARKS_SUCCESS';
export const BOOKMARKS_FAILURE = 'BOOKMARKS_FAILURE';
export const BOOKMARKS_FORM_SUBMITTED = 'BOOKMARKS_FORM_SUBMITTED';

export const BOOKMARKS_MODAL = 'bookmarksModal';
