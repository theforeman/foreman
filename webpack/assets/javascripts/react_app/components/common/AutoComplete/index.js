import React, { Component } from 'react';
import Downshift from 'downshift';
import { Dropdown, MenuItem, InputGroup, FormControl, Button } from 'patternfly-react';
import PropTypes from 'prop-types';

export const renderItems = ({
  items, highlightedIndex, selectedItem, getItemProps,
}) => (
  <Dropdown.Menu
    style={{
      display: 'block',
      position: 'absolute',
      left: 'inherit',
      marginTop: 0,
      top: 'auto',
    }}
  >
    {items.map((item, index) => (
      <MenuItem
        {...getItemProps({
          index,
          item,
          active: highlightedIndex === index,
          onClick: (e) => {
            // At this point the event.defaultPrevented
            // is already set to true by react-bootstrap
            // MenuItem. We need to set it back to false
            // So downshift will execute it's own handler
            e.defaultPrevented = false;
          },
        })}
        key={item}
      >
        {item}
      </MenuItem>
    ))}
  </Dropdown.Menu>
);

class AutoComplete extends Component {
  constructor(props) {
    super(props);

    this.state = {
      inputValue: '',
    };
  }

  render() {
    const {
      onSearch, labelText, onInputUpdate, items, actionText, ...rest
    } = this.props;

    return (
      <Downshift
        onStateChange={({ inputValue }) => {
          if (typeof inputValue === 'string') {
            onInputUpdate(inputValue);
            this.setState({ inputValue });
          }
        }}
        defaultHighlightedIndex={0}
        selectedItem={this.state.inputValue}
        {...rest}
      >
        {({
          getInputProps,
          getItemProps,
          getLabelProps,
          isOpen,
          inputValue,
          highlightedIndex,
          getRootProps,
          selectedItem,
        }) => (
          <div>
            {labelText && <label {...getLabelProps()}>{labelText}</label>}
            <InputGroup>
              <FormControl type="text" {...getInputProps()} />
              <InputGroup.Button>
                <Button onClick={() => onSearch(inputValue)}>{actionText}</Button>
              </InputGroup.Button>
            </InputGroup>

            {isOpen && items.length
              ? renderItems({
                  items,
                  highlightedIndex,
                  selectedItem,
                  getItemProps,
                })
              : null}
          </div>
        )}
      </Downshift>
    );
  }
}

AutoComplete.propTypes = {
  items: PropTypes.arrayOf(PropTypes.string).isRequired,
  onInputUpdate: PropTypes.func.isRequired,
  onSearch: PropTypes.func.isRequired,
  labelText: PropTypes.string,
  actionText: PropTypes.string,
};

AutoComplete.defaultProps = {
  labelText: '',
  actionText: 'Search',
};

export default AutoComplete;
