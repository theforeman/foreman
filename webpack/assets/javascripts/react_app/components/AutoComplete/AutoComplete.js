import React from 'react';
import PropTypes from 'prop-types';
import { TypeAheadSelect } from 'patternfly-react';
import classNames from 'classnames';
import Immutable from 'seamless-immutable';
import { bindMethods, debounceMethods, noop } from '../../common/helpers';
import AutoCompleteMenu from './components/AutoCompleteMenu';
import AutoCompleteSearchButton from './components/AutoCompleteSearchButton';
import AutoCompleteError from './components/AutoCompleteError';
import AutoCompleteAux from './components/AutoCompleteAux';
import AutoCompleteFocusShortcut from './components/AutoCompleteFocusShortcut';
import { STATUS } from '../../constants';
import { TRIGGERS } from './AutoCompleteConstants';
import { KEYCODES } from '../../common/keyCodes';
import { translate as __ } from '../../common/I18n';
import './auto-complete.scss';

class AutoComplete extends React.Component {
  constructor(props) {
    super(props);
    bindMethods(this, [
      'handleClear',
      'handleInputChange',
      'handleResultsChange',
      'handleInputFocus',
      'getResults',
      'windowKeyPressHandler',
      'handleKeyDown',
    ]);
    this._typeahead = React.createRef();
    debounceMethods(this, 500, ['handleLoading']);
  }

  componentDidMount() {
    window.addEventListener('keypress', this.windowKeyPressHandler);
    const {
      controller,
      searchQuery,
      isDisabled,
      error,
      id,
      url,
      initialUpdate,
    } = this.props;

    initialUpdate({ searchQuery, controller, id, isDisabled, error, url });
  }

  componentDidUpdate(prevProps) {
    this.handleLoading();
    const { searchQuery, trigger } = this.props;
    const { RESET, CONTROLLER_CHANGED } = TRIGGERS;
    if (trigger === RESET || trigger === CONTROLLER_CHANGED) {
      this.handleClear();
    }
    if (prevProps.searchQuery !== searchQuery) {
      const typeahead = this._typeahead && this._typeahead.current;
      typeahead && typeahead.setState({ text: searchQuery });
    }
  }

  windowKeyPressHandler(e) {
    const { useKeyShortcuts, handleSearch } = this.props;
    const instance = this._typeahead.current.getInstance();
    const { ENTER, FWD_SLASH, BACK_SLASH } = KEYCODES;
    const { tagName } = e.target;
    const didEventCameFromInput = tagName === 'INPUT' || tagName === 'TEXTAREA';

    /**
     Disable this functionality if the event came from an input,
     or if the 'useKeyShortcuts' is falsy.
    */
    if (didEventCameFromInput || !useKeyShortcuts) {
      return;
    }

    switch (e.charCode) {
      case ENTER: {
        handleSearch();
        break;
      }
      case FWD_SLASH:
      case BACK_SLASH: {
        const {
          focus,
          state: { showMenu },
        } = instance;
        const isMenuHidden = !showMenu;
        if (isMenuHidden) {
          e.preventDefault();
          focus();
        }
        break;
      }
      default: {
        break;
      }
    }
  }

  getResults(searchQuery, trigger, id) {
    const { getResults, controller, url } = this.props;
    getResults({
      url,
      searchQuery,
      controller,
      trigger,
      id,
    });
  }

  handleInputFocus({ target: { value } }) {
    const { id, results } = this.props;
    if (results.length === 0) {
      this.getResults(value, TRIGGERS.INPUT_FOCUS, id);
    }
  }

  handleInputChange(searchQuery) {
    const { id } = this.props;
    this.getResults(searchQuery, TRIGGERS.INPUT_CHANGE, id);
  }

  // Gets the first result from an array of selected results.
  handleResultsChange({ 0: result }) {
    const { id } = this.props;
    if (!result) {
      return;
    }
    this.getResults(result, TRIGGERS.ITEM_SELECT, id);
    /**
     *  HACK: I had no choice but to call to an inner function,
     * due to lack of design in react-bootstrap-typeahead.
     */
    this._typeahead.current.getInstance()._showMenu();
  }

  handleKeyDown({ keyCode }) {
    const instance = this._typeahead.current.getInstance();
    switch (keyCode) {
      case KEYCODES.ENTER: {
        if (!instance.state.activeItem) {
          this.props.handleSearch();
        }
        break;
      }
      case KEYCODES.ESC: {
        instance.blur();
        break;
      }
      default: {
        break;
      }
    }
  }

  handleClear() {
    const { id } = this.props;
    this._typeahead.current.getInstance().clear();
    this.getResults('', TRIGGERS.INPUT_CLEAR, id);
  }

  handleLoading() {
    const { status } = this.props;
    const typeahead = this._typeahead && this._typeahead.current;
    const isLoading = status === STATUS.PENDING;
    typeahead && typeahead.setState({ isLoading });
  }

  componentWillUnmount() {
    window.removeEventListener('keypress', this.windowKeyPressHandler);
    const { resetData, controller, id } = this.props;
    resetData(controller, id);
  }

  render() {
    const {
      id,
      error,
      searchQuery,
      inputProps,
      placeholder,
      results,
      useKeyShortcuts,
      isDisabled,
    } = this.props;
    /** Using a 3rd party library (react-bootstrap-typeahead) that expects a mutable array. */
    const options = Immutable.isImmutable(results)
      ? results.asMutable()
      : results;

    return (
      <div className="foreman-autocomplete">
        <TypeAheadSelect
          id={id}
          ref={this._typeahead}
          defaultInputValue={searchQuery}
          options={options}
          onInputChange={this.handleInputChange}
          onChange={this.handleResultsChange}
          onFocus={this.handleInputFocus}
          onKeyDown={this.handleKeyDown}
          placeholder={__(placeholder)}
          disabled={isDisabled}
          renderMenu={(r, menuProps) => (
            <AutoCompleteMenu {...{ results: r, menuProps }} />
          )}
          inputProps={{
            className: classNames(
              'search-input',
              useKeyShortcuts ? 'use-shortcuts' : ''
            ),
            spellCheck: 'false',
            'data-autocomplete-id': id,
            autoComplete: 'off',
            ...inputProps,
          }}
        />
        <AutoCompleteAux onClear={this.handleClear} />
        <AutoCompleteFocusShortcut useKeyShortcuts={useKeyShortcuts} />
        <AutoCompleteError error={error} />
      </div>
    );
  }
}

AutoComplete.propTypes = {
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  url: PropTypes.string.isRequired,
  results: PropTypes.array,
  searchQuery: PropTypes.string,
  inputProps: PropTypes.object,
  status: PropTypes.string,
  error: PropTypes.string,
  initialError: PropTypes.string,
  controller: PropTypes.string,
  handleSearch: PropTypes.func,
  getResults: PropTypes.func,
  resetData: PropTypes.func,
  initialUpdate: PropTypes.func,
  useKeyShortcuts: PropTypes.bool,
  placeholder: PropTypes.string,
  isDisabled: PropTypes.bool,
  trigger: PropTypes.string,
};

AutoComplete.defaultProps = {
  results: [],
  searchQuery: '',
  inputProps: {},
  status: null,
  error: null,
  initialError: null,
  controller: null,
  handleSearch: noop,
  getResults: noop,
  resetData: noop,
  initialUpdate: noop,
  useKeyShortcuts: false,
  placeholder: 'Filter ...',
  isDisabled: false,
  trigger: null,
};

AutoComplete.SearchButton = AutoCompleteSearchButton;
AutoComplete.Error = AutoCompleteError;

export default AutoComplete;
