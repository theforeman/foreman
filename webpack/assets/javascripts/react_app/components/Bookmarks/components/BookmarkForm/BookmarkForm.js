import React from 'react';
import PropTypes from 'prop-types';
import * as Yup from 'yup';

import { noop } from '../../../../common/helpers';
import ForemanForm from '../../../common/forms/ForemanForm';
import TextField from '../../../common/forms/TextField';
import { translate as __ } from '../../../../../react_app/common/I18n';
import { maxLengthMsg, requiredMsg } from '../../../common/forms/validators';

const bookmarkFormSchema = Yup.object().shape({
  name: Yup.string()
    .max(...maxLengthMsg(254))
    .required(requiredMsg()),
  query: Yup.string()
    .max(...maxLengthMsg(4096))
    .required(requiredMsg()),
});

const BookmarkForm = ({
  url,
  submitForm,
  controller,
  onCancel,
  initialValues,
  setModalClosed,
}) => {
  const handleSubmit = async (values, actions) => {
    await submitForm({
      url,
      values: { ...values, controller },
      item: 'Bookmarks',
      message: __('Bookmark was successfully created.'),
    });
    setModalClosed();
  };

  return (
    <ForemanForm
      onSubmit={(values, actions) => handleSubmit(values, actions)}
      initialValues={initialValues}
      validationSchema={bookmarkFormSchema}
      onCancel={onCancel}
    >
      <TextField name="name" type="text" required="true" label={__('Name')} />
      <TextField
        name="query"
        type="textarea"
        required="true"
        label={__('Query')}
        inputClassName="col-md-8"
      />
      <TextField name="public" type="checkbox" label={__('Public')} />
    </ForemanForm>
  );
};

BookmarkForm.propTypes = {
  onCancel: PropTypes.func,
  submitForm: PropTypes.func.isRequired,
  controller: PropTypes.string.isRequired,
  initialValues: PropTypes.object.isRequired,
  url: PropTypes.string.isRequired,
  setModalClosed: PropTypes.func.isRequired,
};

BookmarkForm.defaultProps = {
  onCancel: noop,
};

export default BookmarkForm;
