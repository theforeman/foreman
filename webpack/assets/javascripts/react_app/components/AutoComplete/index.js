import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import * as actions from './AutoCompleteActions';
import reducer from './AutoCompleteReducer';
import AutoComplete from './AutoComplete';
import {
  selectAutocompleteError,
  selectAutocompleteResults,
  selectAutocompleteSearchQuery,
  selectAutocompleteStatus,
  selectAutocompleteIsDisabled,
  selectAutocompleteUrl,
  selectAutocompleteTrigger,
} from './AutoCompleteSelectors';

const mapStateToProps = (state, ownProps) => {
  const { id } = ownProps;
  return {
    error: selectAutocompleteError(state, id, ownProps),
    results: selectAutocompleteResults(state, id, ownProps),
    searchQuery: selectAutocompleteSearchQuery(state, id, ownProps),
    status: selectAutocompleteStatus(state, id, ownProps),
    disabled: selectAutocompleteIsDisabled(state, id, ownProps),
    url: selectAutocompleteUrl(state, id, ownProps),
    trigger: selectAutocompleteTrigger(state, id, ownProps),
  };
};

const mapDispatchToProps = dispatch => bindActionCreators(actions, dispatch);

export const reducers = { autocomplete: reducer };

export default connect(mapStateToProps, mapDispatchToProps)(AutoComplete);
