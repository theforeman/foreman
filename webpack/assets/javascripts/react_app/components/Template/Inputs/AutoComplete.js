import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { FieldLevelHelp } from 'patternfly-react';

import Form from '../../common/forms/CommonForm';
import AutoComplete from '../../AutoComplete';

const TemplateAutoComplete = ({
  label,
  resourceType,
  id,
  searchQuery,
  initialError,
  info,
  isRequired,
  url,
  template,
}) => (
  <Form
    touched
    label={label}
    error={initialError}
    required={isRequired}
    tooltipHelp={
      info && (
        <FieldLevelHelp
          buttonClass="field-help"
          content={<Fragment>{info}</Fragment>}
        />
      )
    }
  >
    <AutoComplete
      id={`template-autocomplete-input-${id}`}
      inputProps={{
        name: `${template}[input_values][${id}][value]`,
      }}
      controller={resourceType}
      url={url}
      searchQuery={searchQuery}
    />
  </Form>
);
TemplateAutoComplete.propTypes = {
  resourceType: PropTypes.string,
  searchQuery: PropTypes.string,
  initialError: PropTypes.string,
  label: PropTypes.string.isRequired,
  info: PropTypes.string,
  isRequired: PropTypes.bool,
  id: PropTypes.number.isRequired,
  url: PropTypes.string.isRequired,
  template: PropTypes.string.isRequired,
};

TemplateAutoComplete.defaultProps = {
  resourceType: null,
  searchQuery: '',
  initialError: null,
  info: null,
  isRequired: false,
};

export default TemplateAutoComplete;
