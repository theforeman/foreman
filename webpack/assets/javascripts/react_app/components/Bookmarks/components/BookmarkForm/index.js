import { connect } from 'react-redux';
import BookmarkForm from './BookmarkForm';
import { submitForm } from '../../../../redux/actions/common/forms';
import { selectAutocompleteSearchQuery } from '../../../AutoComplete/AutoCompleteSelectors';

const mapStateToProps = (state, { controller }) => ({
  initialValues: {
    public: true,
    query: selectAutocompleteSearchQuery(state, 'searchBar', { controller }),
  },
});

const mapDispatchToProps = {
  submitForm,
};

export default connect(mapStateToProps, mapDispatchToProps)(BookmarkForm);
