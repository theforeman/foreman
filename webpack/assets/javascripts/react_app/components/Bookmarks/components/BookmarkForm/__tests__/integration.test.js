import React from 'react';
import uuidV1 from 'uuid/v1';
import { IntegrationTestHelper } from '@theforeman/test';
import API from '../../../../../redux/API/API';
import { APIMiddleware } from '../../../../../redux/API';

import BookmarkForm from '../index';
import { reducers as bookmarksReducer } from '../../../index';
import { reducers as autocompleteReducer } from '../../../../AutoComplete/index';
import foremanModalsReducer from '../../../../ForemanModal/ForemanModalReducer';
import {
  response,
  name,
  search,
  publik,
  item,
  submitResponse,
  controller,
} from '../../../Bookmarks.fixtures';
import { BOOKMARKS_SUCCESS } from '../../../BookmarksConstants';

const reducers = {
  foremanModals: foremanModalsReducer,
  ...bookmarksReducer,
  ...autocompleteReducer,
};

jest.mock('../../../../../redux/API/API');
jest.mock('uuid/v1');
uuidV1.mockImplementation(() => '1547e1c0-309a-11e9-98f5-5f761412a4c2');

const props = {
  url: '/api/v2/hosts',
  controller: 'hosts',
  setModalClosed: jest.fn(),
};

describe('Bookmark form integration test', () => {
  it('should allow submission when fields filled in', async () => {
    API.post.mockImplementation(async () => submitResponse);

    const testHelper = new IntegrationTestHelper(reducers, [APIMiddleware]);
    testHelper.store.dispatch({
      type: BOOKMARKS_SUCCESS,
      payload: {
        controller,
        item,
      },
      response: response.data,
    });

    const component = testHelper.mount(<BookmarkForm {...props} />);

    expect(
      component.find('Button[bsStyle="primary"]').props().disabled
    ).toBeTruthy();
    expect(
      component.find('Button[bsStyle="default"]').props().disabled
    ).not.toBeTruthy();
    component
      .find('input[name="name"]')
      .simulate('change', { target: { name: 'name', value: name } });
    component
      .find('textarea[name="query"]')
      .simulate('change', { target: { name: 'query', value: search } });
    component
      .find('input[name="public"]')
      .simulate('change', { target: { name: 'public', value: publik } });

    const submitBtn = component.find('Button[bsStyle="primary"]');

    expect(submitBtn.disabled).not.toBeTruthy();
    expect(
      component.find('Button[bsStyle="default"]').props().disabled
    ).not.toBeTruthy();

    submitBtn.simulate('submit');
    await IntegrationTestHelper.flushAllPromises();
    component.update();

    testHelper.takeActionsSnapshot('form submitted');
  });
});
