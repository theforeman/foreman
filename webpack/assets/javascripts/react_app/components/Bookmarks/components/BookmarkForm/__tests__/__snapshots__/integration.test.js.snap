// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bookmark form integration test should allow submission when fields filled in: form submitted 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "controller": "hosts",
        "item": "Bookmark",
      },
      "response": Object {
        "results": Array [
          Object {
            "controller": "hosts",
            "id": 52,
            "name": "1111",
            "owner_id": 1,
            "owner_type": "User",
            "public": true,
            "query": "abc",
          },
          Object {
            "controller": "hosts",
            "id": 54,
            "name": "1122",
            "owner_id": 1,
            "owner_type": "User",
            "public": true,
            "query": "abc",
          },
        ],
      },
      "type": "BOOKMARKS_SUCCESS",
    },
  ],
  Array [
    Object {
      "key": "BOOKMARKS_FORM_SUBMITTED",
      "payload": Object {
        "url": "/api/v2/hosts",
      },
      "type": "BOOKMARKS_FORM_SUBMITTED_REQUEST",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "actionTypes": undefined,
        "errorToast": [Function],
        "handleError": [Function],
        "handleSuccess": [Function],
        "headers": undefined,
        "key": "BOOKMARKS_FORM_SUBMITTED",
        "params": Object {
          "controller": "hosts",
          "name": "Joe.D",
          "public": false,
          "query": "name ~ my",
        },
        "successToast": [Function],
        "url": "/api/v2/hosts",
      },
      "type": "API_POST",
    },
  ],
  Array [
    Object {
      "key": "BOOKMARKS_FORM_SUBMITTED",
      "payload": Object {
        "url": "/api/v2/hosts",
      },
      "response": Object {
        "controller": "hosts",
        "name": "Joe.D",
        "public": false,
        "query": "name ~ my",
      },
      "type": "BOOKMARKS_FORM_SUBMITTED_SUCCESS",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "key": "BOOKMARKS_FORM_SUBMITTED_SUCCESS",
        "message": Object {
          "key": "BOOKMARKS_FORM_SUBMITTED_SUCCESS",
          "message": "Bookmark was successfully created.",
          "type": "success",
        },
      },
      "type": "TOASTS_ADD",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "data": Object {
          "controller": "hosts",
          "name": "Joe.D",
          "public": false,
          "query": "name ~ my",
        },
        "item": "Bookmarks",
      },
      "type": "BOOKMARKS_FORM_SUBMITTED",
    },
  ],
]
`;
